File: MastersProject/getLMiPod.py
import sys
import os
import numpy as np
from numpy import mean, sqrt, square
from utilities import  rms, find


"""
 The LM matrix is made based on paramsiPod strucutre, RMS values, and
 up2Down1 vlaues.
 
 Input
 paramsiPod -input structure that define criteria for scoring (The WASM criteria
             for scoring different leg movements)
 RMS - root mean value square of accelometer data.
 up2Down1 - showing if the patient whether the patient is in sleep or awake.

 Output
 LM - matrix of leg movement having minimum duration of 0.5 seconds and no 
      max duration. The structure include differnet type of variables.
"""
def getLMiPod(paramsiPod,RMS,up2Down1):
    #print("start of: getLMiPod")
    if RMS[0] == None:
        LM = None
        return
    LM = findIndices(RMS,paramsiPod.lowThreshold,paramsiPod.highThreshold,paramsiPod.minLowDuration,paramsiPod.minHighDuration,paramsiPod.fs)
    
    if LM.size == 0:
        return
    
    #%% Median must pass lowthreshold
    if paramsiPod.morphologyCriteria == 'on':
        LM = cutLowMedian(RMS,LM,paramsiPod.lowThreshold,paramsiPod.fs)
    
    #%% Add Duration in 3rd column
    LM = np.insert(LM, 2, values = (LM[:,1]-LM[:,0])/paramsiPod.fs, axis = 1)

    #Add IMI in sec in 4th column
    col_4=[]
    col_4.append(9999)
    col_4.extend((LM[1:LM.shape[0],0] - LM[0:LM.shape[0]-1,0])/paramsiPod.fs)
    
    if LM.shape[0] > 1:
        LM = np.insert(LM, 3, values = col_4, axis = 1)

    #5th column as zeros (reserved for PLM)
    LM = np.insert(LM,4,values=np.empty(LM.shape[0]),axis=1)
    LM_i = np.array(LM, dtype='i')

    # Add Up2Down1 in 6th Column (5th col reserved for PLM)
    LM = np.insert(LM,5,values = up2Down1[LM_i[:,0],0], axis = 1)
    LM_i = np.array(LM, dtype='i')

    # Convert beginning of LMs to mins in 7th Column
    LM = np.insert(LM,6,values = LM[:,0]/(paramsiPod.fs*60), axis=1)
    LM_i = np.array(LM, dtype='i')
    
    # Record epoch number of LM in 8th Column
    LM = np.insert(LM,7,values = np.round(LM[:,6]*2 + 0.5),axis=1)
    LM_i = np.array(LM, dtype='i')
    #print("LM 8th col:",LM)

    # Record breakpoints after long LM in 9th column
    col_9=[]
    col_9.append(1)
    col_9.extend(LM[0:LM_i.shape[0]-1,2] > paramsiPod.maxCLMDuration)
    LM = np.insert(LM,8,values=col_9,axis=1)
 

    LM_i = np.array(LM, dtype='i')
    # Record area of LM in 10th Column 
    col_10 = []
    for i in range(LM.shape[0]):
        col_10.append(np.sum(RMS[LM_i[i,0]:LM_i[i,1]])/paramsiPod.fs)
    LM = np.insert(LM,9,values=col_10,axis=1)

    return LM
#########################################################


#Remove leg movements whose median activity is less than noise level
def cutLowMedian(dsEMG,LM1,min,fs,**kwargs):
    LM1 = LM1.astype(float) #np.array(LM1, dtype='f')
    opt = kwargs.get('opt', 1)
    if opt:
        opt = 1
    
    # Calculate duration, probably don't need this though
    LM1 = np.insert(LM1,2,values = (LM1[:,1]-LM1[:,0])/fs,axis=1)

    # Add column with median values
    LM_i = np.array(LM1, dtype='i')
    median_col=[]
    for i in range(LM1.shape[0]):
        temp = np.median(dsEMG[LM_i[i,0]:LM_i[i,1]])
        median_col.append(temp)
    LM1 = np.insert(LM1,3,values=median_col,axis=1)

    if opt == 1:
        LM = shrinkWindow(LM1,dsEMG,fs,min)
    elif opt == 2:
        LM = tryShrinking(LM1,dsEMG,fs,min)
    elif opt == 3:
        LM = LM1


    # Exclude movements that still are empty
    LM = LM[LM[:,3] > min,0:2]
    return LM
    
#A different method of checking for a movement within the movement. This
# searches for any 0.5 second window where the median is above noise level.
def shrinkWindow(LM,dsEMG,fs,min):
    
    empty = find(LM[:,3], lambda x: x < min)
    #print("empty ",empty)
    for i in range(len(empty)):
        initstart = int(LM[empty[i],0])
        initstop = int(LM[empty[i],1])
        a = np.median(dsEMG[initstart:initstop])
        start = int(LM[empty[i]][0])
        stop = int(start + fs/2)
        while a < min and stop < initstop:
            a =  np.median(dsEMG[start:stop])
            start = int(start + fs/10)
            stop = int(start + fs/2)
        LM[empty[i],3] = a
    return LM

#Attempt to reduce the duration of the event in order to find a movement
#that fits minimum density requirement.
def tryShrinking(LM1,dsEMG,fs,min):
    short = find(LM1[:,3], lambda x: x < min)
    for i in range(len(short)):
        start = int(LM1[short[i]][0])
        stop = int(LM1[short[i]][1])
        while (start - stop)/fs > 0.6 and np.median(dsEMG[start:stop]) < min:
            stop = stop - fs/10
        LM1[short[i],3] = np.median(dsEMG[start:stop])
    return short

def findIndices(data,lowThreshold,highThreshold,minLowDuration,minHighDuration,fs):
    fullRuns = np.empty([2,2])# [[-1 for x in range(2)] for y in range(2)] 
    minLowDuration = minLowDuration * fs
    minHighDuration = minHighDuration * fs
    lowValues = find(data, lambda x: x < lowThreshold) 
    highValues = find(data, lambda x: x > highThreshold)

    if len(highValues) < 1:
        fullRuns[0][0] = 0
        fullRuns[0][1] = 0
    elif len(lowValues) < 1:
        fullRuns[0][0] = 1
        fullRuns[0][1] = 0

    lowRuns = returnRuns(lowValues,minLowDuration)
    highRuns = []
    numHighRuns = 0
    searchIndex = highValues[0]
    while searchIndex < data.shape[0]:
        distToNextLowRun,lengthOfNextLowRun = calcDistToRun(lowRuns,searchIndex)
        if distToNextLowRun == -1: ## Then we have hit the end, record our data and stop 
            highRuns.append([searchIndex , data.shape[0]])
        else: ##We have hit another low point, so record our data, 
            highRuns.append([searchIndex , searchIndex + distToNextLowRun-1])            
            searchIndex = searchIndex + distToNextLowRun + lengthOfNextLowRun
            highValues = np.asarray(highValues)
            temp = np.argwhere(highValues > searchIndex)
            if temp.size != 0:
                searchIndex = highValues[int(temp[0])]
        numHighRuns = numHighRuns + 1
    #Implement a quality control to only keep highRuns > minHighDuration
    highRuns = np.array(highRuns)
    runLengths = highRuns[:,1]-highRuns[:,0]
    ind = find(runLengths, lambda x: x > minHighDuration)
    fullRuns = highRuns[ind]
    #print("fullRuns ",fullRuns)
    return fullRuns




def returnRuns(vals,duration): 
    vals = np.asarray(vals)
    k = (np.diff(vals) != 1).astype(int)
    k = np.insert(k, 0, 1, axis=0)
    s = np.cumsum(k)
    x = np.histogram(s,np.arange(1,s[-1]+2))[0]
    idx = find(k, lambda x: x != 0) 
    idx = np.asarray(idx)
    startIndices = vals[idx[x>=duration]]
    stopIndices = startIndices + x[x>=duration] - 1
    runs = [startIndices,stopIndices]
    return runs

"""
 This Function finds the closest next (sequential) run from the current
 position.  It does include prior runs.  If there is no run it returns a
 distance of -1, otherwise it returns the distance to the next run.  It
 will also return the length of that run.  If there is no run it returns a
 length of -1.
 """
def calcDistToRun(run,position):
    distList = run[:][0] - position
    distPos = distList[distList>0]
    if distPos.all():
        dist = min(distPos)
        runIndex = int(np.where(distList == dist)[0])
        length = run[1][runIndex] - run[0][runIndex] + 1
    else:
        length = -1
        dist = -1
    return dist,length

File: matlabcode/subroutines/getLMiPod.m
function LM = getLMiPod(paramsiPod,RMS,up2Down1)
% The LM matrix is made based on paramsiPod strucutre, RMS values, and
% up2Down1 vlaues.
% 
% Input
% paramsiPod -input structure that define criteria for scoring (The WASM criteria
%             for scoring different leg movements)
% RMS - root mean value square of accelometer data.
% up2Down1 - showing if the patient whether the patient is in sleep or awake.
%
% Output
% LM - matrix of leg movement having minimum duration of 0.5 seconds and no 
%      max duration. The structure include differnet type of variables.

if isnan(RMS(1,1))
    LM=NaN;
    return;
end
   
LM=findIndices(RMS,paramsiPod.lowThreshold,paramsiPod.highThreshold,paramsiPod.minLowDuration,paramsiPod.minHighDuration,paramsiPod.fs);
if isempty(LM)
   return; 
end
%% Median must pass lowthreshold
if strcmp(paramsiPod.morphologyCriteria,'on')
    LM = cutLowMedian(RMS,LM,paramsiPod.lowThreshold,paramsiPod.fs);
end
%% Add Duration in 3rd column
LM(:,3)=(LM(:,2)-LM(:,1))/paramsiPod.fs;
%% Add IMI in sec in 4th column
LM(1,4)=9999;
if (size(LM,1)>1)
    LM(2:size(LM,1),4)=(LM(2:size(LM,1),1)-LM(1:size(LM,1)-1,1))/paramsiPod.fs;
end
%% Add Up2Down1 in 6th Column (5th col reserved for PLM)
LM(:,6)=up2Down1(LM(:,1),1);
%% Convert beginning of LMs to mins in 7th Column
LM (:,7) = LM(:,1)/(paramsiPod.fs*60);
%% Record epoch number of LM in 8th Column
LM (:,8) = round (LM (:,7)*2 +0.5);
%% Record breakpoints after long LM in 9th column
LM(:,9)=[1;LM(1:end-1,3)>paramsiPod.maxCLMDuration];
%% Record area of LM in 10th Column 
for i=1:size(LM,1)
    LM(i,10)=sum(RMS(LM(i,1):LM(i,2),1))/paramsiPod.fs;
end

end







%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% 
%getLM Subfunctions
%% Cut low median
function [LM] = cutLowMedian(dsEMG,LM1,min,fs,opt)
%% Remove leg movements whose median activity is less than noise level
% [LM] = cutLowMedian(dsEMG,LM1,min,fs,opt)
%
% cutLowMedian can either remove movements immediately or attempt to find a
% subsection with median above noise.
%
% Inputs: 
%   dsEMG - filtered and rectified EMG signal
%   LM1 - the initial LM array, with at least start and stop times
%   min - the minimum value (noise level) that medians must pass
%   fs - sampling rate
%   opt - (1) moving 1/2 second window (2) shrinking window, starting at
%       the original bounds of the movement(3) immediately discard. If no
%       option is selected, default is (1).
%
% Outputs:
%   LM - the new array with movements that did not pass the median test
%       removed

if ~exist('opt','var')
    opt = 1;
end

% Calculate duration, probably don't need this though
LM1(:,3) = (LM1(:,2) - LM1(:,1))/fs;

% Add column with median values
for i = 1:size(LM1,1)
    LM1(i,4) = median(dsEMG(LM1(i,1):LM1(i,2)));
end

%tic

switch (opt)
case 1
    LM = shrinkWindow(LM1,dsEMG,fs,min);
case 2
    LM = tryShrinking(LM1,dsEMG,fs,min);
case 3
    LM = LM1;
end

%fprintf('It took %.2f seconds to shrink this\n',toc);

% Exclude movements that still are empty
LM = LM(LM(:,4) > min,1:2);
end

% Attempt to reduce the duration of the event in order to find a movement
% that fits minimum density requirement.

function LM1 = tryShrinking(LM1,dsEMG,fs,min)
short = find(LM1(:,4) < min);

med = @(a,b) median(dsEMG(a:b)); % macro to get median

for i = 1:size(short,1)
    
    start = LM1(short(i),1); stop = LM1(short(i),2);
    
    % Continuously reduce the stop time of the movement and check the
    % median again to see if it passes the minimum
    while (stop - start)/fs > 0.6 && med(start,stop) < min
        stop = stop - fs/10;
    end
    
    % Mark the new median in the original array
    LM1(short(i),4) = med(start,stop);
end
end


% A different method of checking for a movement within the movement. This
% searches for any 0.5 second window where the median is above noise level.
function LM = shrinkWindow(LM,dsEMG,fs,min)

empty = find(LM(:,4) < min);

med = @(a,b) median(dsEMG(a:b)); % macro to get median

for i = 1:size(empty,1)
    
    % Record the original start, stop and median
    initstart = LM(empty(i),1);
    initstop = LM(empty(i),2);
    a = med(initstart,initstop);
    
    % Now, the start and stop times that will shift
    start = LM(empty(i),1); stop = start + fs/2;
    
    % Loop through 1/2 second windows of the movement, skipping along at
    % 1/10 second.
    while a < min && stop < initstop;
        a = med(start,stop);
        start = start + fs/10; stop = start + fs/2;
    end
    
    % Mark the new median in the original array
    LM(empty(i),4) = a;
end

end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%findIndices

%%  [Indices] = findIndices(data,lowThreshold,highThreshold,minLowDuration,minHighDuration)
%%  returns array of indices defining the periods of data above the highThreshold that last for
%%  a minimum length of minHighDuration, without an interruption below lowThreshold that lasts for
%%  a minimum length of minLow Duration,
%%%%minHighDuration
function [fullRuns] = findIndices(data,lowThreshold,highThreshold,minLowDuration,minHighDuration,fs)
    minLowDuration = minLowDuration * fs;
    minHighDuration = minHighDuration * fs;
    lowValues = find(data < lowThreshold);          
    highValues = find(data > highThreshold);
    if size(highValues,1) < 1
        fullRuns(1,1) = 0;                  %ends function if no highvalues detected 
        fullRuns(1,2) = 0                   %sets array with runs to  0  0               
        return;
    end
    if size(lowValues,1) < 1
        fullRuns(1,1) = 1;                  %ends function if not lowValues detected
        fullRuns(1,2) = 0                   % sets array with runs to 1  0
        return;
    end 

    lowRuns = returnRuns(lowValues,minLowDuration);
    %highRuns = returnRuns(highValues,minHighDuration);

    numHighRuns = 0;
    searchIndex = highValues(1);
    while (searchIndex < size(data,1))
        numHighRuns = numHighRuns + 1;
        [distToNextLowRun,lengthOfNextLowRun] = calcDistToRun(lowRuns,searchIndex);
        if (distToNextLowRun == -1)  %%Then we have hit the end, record our data and stop 
            highRuns(numHighRuns,1) = searchIndex;
            highRuns(numHighRuns,2) = size(data,1);
            searchIndex = size(data,1);
        else %%We have hit another low point, so record our data, 
            highRuns(numHighRuns,1) = searchIndex;
            highRuns(numHighRuns,2) = searchIndex + distToNextLowRun-1;
            %%And then search again with the next high value after this low Run.
            searchIndex = searchIndex+distToNextLowRun+lengthOfNextLowRun;
            searchIndex = highValues(find(highValues>searchIndex,1,'first'));
        end
    end
      
    %%Implement a quality control to only keep highRuns > minHighDuration
    runLengths = highRuns(:,2)-highRuns(:,1);
    fullRuns = highRuns(find(runLengths > minHighDuration),:);
    
    
    
    
        
    
    %%%CREATE GRAPHICAL OUTPUT SO THAT I CAN SEE THAT IT"S WORKING
    
%     Indices = highRuns;
%     figure;
%     hold on;
%     plot(data,'b')
%     plot([1,size(data,1)],[lowThreshold,lowThreshold],'r--');
%     plot([1,size(data,1)],[highThreshold,highThreshold],'r--');
%     for i=1:size(Indices,1)
%         plot([Indices(i,1),Indices(i,2)],[highThreshold,highThreshold],'g-','LineWidth',5);
%     end
%     
%     Indices = lowRuns;
%     for i=1:size(Indices,1)
%         plot([Indices(i,1),Indices(i,2)],[lowThreshold,lowThreshold],'m-','LineWidth',5);
%     end
%     
%     Indices = fullRuns;
%     for i=1:size(Indices,1)
%         plot([Indices(i,1),Indices(i,2)],[mean([highThreshold,lowThreshold]),mean([highThreshold,lowThreshold])],'r-','LineWidth',5);
%     end
%     

end





function [runs] = returnRuns(vals,duration)
    k = [true;diff(vals(:))~=1 ];
    s = cumsum(k);
    x =  histc(s,1:s(end));
    idx = find(k);
    startIndices = vals(idx(x>=duration));
    stopIndices = startIndices + x(x>=duration) -1;
    runs =  [startIndices,stopIndices];
end



%%This Function finds the closest next (sequential) run from the current
%%position.  It does include prior runs.  If there is no run it returns a
%%distance of -1, otherwise it returns the distance to the next run.  It
%%will also return the length of that run.  If there is no run it returns a
%%length of -1.
function [dist,length] = calcDistToRun(run,position)
    distList = run(:,1) - position;
    distPos = distList(distList > 0);
    if (distPos) 
        dist = min(distPos);
        runIndex = find(distList == dist);
        length = run(runIndex,2)-run(runIndex,1)+1;
    else
        length = -1;
        dist = -1;
    end
end